### PR TITLE
feat: build envs on install/activate

### PR DESCRIPTION
**Pull request by zmitchell**
## Proposed Changes

Attempt to transactionally build the environment when installing, uninstalling, or activating. This has the added benefit of ensuring that you can't install a package that doesn't exist.

## Release Notes

Installing/uninstalling a package now actually builds the environment rather than simply putting the package name in the manifest.

### CODE REVIEW

File: crates/flox-rust-sdk/src/models/environment/path_environment.rs
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/path_environment.rs:203]: question blocking: dropping `_nix` (and the earlier decision to not use `_nix`) means that the `nix` being used to build the environment doesn't have any of our configuration, e.g. `NIX_SSL_CERT_FILE`. How are we planning to ensure the `nix` we use here is properly configured? Or is that being passed somehow or do we know we don't need to pass config?

File: tests/environment-install.bats
Comment (mkenigs) [inline tests/environment-install.bats:80]: suggestion blocking: add a test

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: crates/flox-rust-sdk/src/models/environment/path_environment.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment/path_environment.rs:203]: I think that's something that should be handled on the builder side (`build-env.sh`). I'm not sure what all configuration is required, but I'd be happy to add it to the builder script.

File: crates/flox-rust-sdk/src/models/environment/path_environment.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment/path_environment.rs:268]: Ah, I forgot that was already happening in `transact_with_manifest_contents`

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment/managed_environment.rs:93]: Yeah I just missed that

File: crates/flox-rust-sdk/src/models/environment/path_environment.rs
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/environment/path_environment.rs:203]: I'm not sure if some of it's cruft, but I think this is where we're setting stuff up:
https://github.com/flox/flox/blob/93785d22b78e59e7a5dcaa4cdf1e18956c608e50/crates/flox-rust-sdk/src/flox.rs#L474

Since that's something we want for the entirety of the CLI, I don't think it makes sense to move it to `build-env.sh`. Then if we want to invoke `nix` to do anything else we have to duplicate the config. It's possible we could at some point let `pkgdb` handle all the config but I'm not sure we're to that point yet

Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/environment/path_environment.rs:203]: I also don't think we need to fix setting all the config in this PR, but I'd rather not drop code for the prior method of setting config without a clear plan for what we'll do instead

File: tests/environment-install.bats
Comment (zmitchell) [reply tests/environment-install.bats:80]: I'm confused, we already have tests for this. I made the change here specifically because the test fails with package `foo`, which doesn't exist.

Comment (mkenigs) [reply tests/environment-install.bats:80]: The existing tests only check that the manifest gets modified. You could check:
1. Installing an invalid package fails with a helpful message
2. `$PROJECT_DIR/.flox/run/$system.$name/bin` contains `hello`

File: crates/flox-rust-sdk/src/models/environment/path_environment.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment/path_environment.rs:203]: As far as I can tell, this is the only real usage of the Nix CLI left in `flox`. There's a handful of Nix CLI calls that are in the codebase, but I don't think they're actually codepaths we use today. We also aren't dropping any code for the prior method, it still exists.

It's hard to say what the plan for Nix config should look like going forward without knowing what the plan is for what the next builder will look like. If I remember correctly @aakropotkin suggested that the next builder may not call the Nix CLI at all, in which case the config could be something that we pass along in a JSON blob to whatever this builder is.

Setting all the flags that we set in `build-env.sh` requires a PR to `runix` because it doesn't support all of those flags for `nix build` yet (`--file`, `--print-out-path`, `--arg`, `--argstr`). Can we merge this PR, make the update to `runix`, and then come back here to replace `build-env.sh` entirely with a call to the `Build` struct until we figure out a more coherent plan?

Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/environment/path_environment.rs:203]: Calling `build.run()` takes `nix` as an argument, so you'll have to revert all the places you drop `nix` in this PR, which is what I meant by dropping code.

Resolving this thread as I wanted to raise/discuss the issue, but I don't think it needs to block this code from getting in.

Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment/path_environment.rs:203]: It's a good callout, thanks for reminding me. I totally forgot that we have a bunch of config for Nix calls.

> Calling build.run() takes nix as an argument

Yes, but you can get a `nix` via `flox.nix()`, so the interface doesn't need to change.

File: crates/flox-rust-sdk/src/models/environment/mod.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/environment/mod.rs:212]: thanks for those, I've been meaning to add these for a while now

File: crates/flox-rust-sdk/src/models/environment/mod.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/environment/mod.rs:362]: question (non blocking here): do you know if we can already pass lockfile content here too?

that would make building a bit more convenient for managed envs where we dont have direct access to the files.

File: crates/flox-rust-sdk/src/models/environment/mod.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/environment/mod.rs:366]:
```suggestion
    debug!(target: "posix", "locking manifest with command: {pkgdb_cmd:?}");
```

for consistencies sake, can we leave them on the "posix" target for now, or remove the target from the git one as well?

File: crates/flox-rust-sdk/src/models/environment/mod.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/environment/mod.rs:377]: I guess should be general output on stdout and **error**\* messages on stderr?
\* errors since its non-successful not because its std*err*

Not sure what exactly will be on these actually though..

```suggestion
        debug!(
            "pkgdb stdout content: {}",
            String::from_utf8_lossy(&output.stdout)
        );
        error!(
            "pkgdb stderr content: {}",
            String::from_utf8_lossy(&output.stderr)
        );
```

File: crates/flox-rust-sdk/src/models/manifest.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/manifest.rs:79]: looks a lot like print debugging :)
if this was intentional to be left in lets at least hide it under debug

```suggestion
    debug!("MANIFEST CONTENTS BEFORE: {}", manifest_contents);
```

File: crates/flox-rust-sdk/src/models/manifest.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/manifest.rs:88]: suggestion (wording, non-blocking)

wouldnt `MissingInstallTable` print something like: "No packages installed"?
sounds a bit more precise than `PackageNotFound`

File: crates/flox-rust-sdk/src/models/manifest.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/manifest.rs:108]: same as above
```suggestion
    debug!("MANIFEST CONTENTS AFTER: {}", toml);
```

File: crates/flox-rust-sdk/src/models/manifest.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/manifest.rs:88]: just saw the tests output below, i guess for single package uninstalls this does look better with the package included 👍🏼

**Review by ysndr — CHANGES_REQUESTED** [review]

File: crates/flox-rust-sdk/src/models/environment/mod.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment/mod.rs:377]: This is actually a mistake, this was left in while I was debugging something

File: crates/flox-rust-sdk/src/models/manifest.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/manifest.rs:79]: Also accidentally left in 😅

Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/manifest.rs:79]: Also accidentally left in 😅

File: crates/flox-rust-sdk/src/models/manifest.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/manifest.rs:108]: Also left in by mistake

File: crates/flox-rust-sdk/src/models/manifest.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/manifest.rs:88]: I actually had a separate `NoPackagesFound` variant as part of an intermediate state of this commit and decided to just go with the existing error variant. It gets the point across that the package you want to uninstall isn't there so you go look at your manifest and see that there's _no_ packages there.

File: crates/flox-rust-sdk/src/models/environment/mod.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment/mod.rs:366]: I'll just remove all of them, they end up being formatted differently for whatever reason. I've never actually seen someone use target like this. Is it meant to address some bug or edge case?

File: crates/flox-rust-sdk/src/models/environment/mod.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment/mod.rs:362]: Yeah, you can pass a file path or inline JSON for this flag now I believe

**Comment by zmitchell** [comment]
I've discovered a bug in the transaction machinery that makes it such that the lockfile from the temporary transaction environment doesn't get copied back to the original environment. Once that's done this will be ready to merge.

**Comment by zmitchell** [comment]
It turned out not to be a bug in the transaction machinery, but instead a bug in the lockfile/manifest discovery. I fixed that and this should be ready to go.

The acceptance test in #464 is still failing, but at this point I can't demonstrate that it's a bug in `flox`, as the lockfile we get back from `pkgdb` as produced by that test has a different version than what the test is expecting. This search failure is also unrelated to these changes, so I would prefer not to hold this PR up because of that and address the search behavior in a separate PR.

**Review by aakropotkin — CHANGES_REQUESTED** [review]
acceptance-criteria blocking: You need to pass that acceptance test for the PR to be accepted and to close the issue.

**Comment by zmitchell** [comment]
This is unrelated to that failure, and the ticket doesn't relate to search at all. The ticket doesn't mention search failures at all, so please don't add acceptance criteria after the fact, especially if the PR is already green.

**Comment by zmitchell** [comment]
I've made it clear that I'm going to continue working on the search failure in a separate PR.

**Comment by aakropotkin** [comment]
Gotcha, I was confused by the slack links about #464.

**Review by mkenigs — APPROVED** [review]

File: crates/flox-rust-sdk/src/models/environment/mod.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment/mod.rs:370]: It was meant so these dont always show up and were formatted differently to be spotted and copied more easily. There was special handling for these kinds of logs in flox-bash which we ended up trying to replicate in the rust version..

File: crates/flox-rust-sdk/src/models/environment/path_environment.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/environment/path_environment.rs:424]: The env-link created during install will not be recorded by nix as a valid gcroot!
once the transaction is done, we should still set a gc root such that the environment is not garbage collected by nix.

Granted, we don't perform gc's with nix at the moment.

Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment/path_environment.rs:424]: Good catch

**Pull request merged**
